### DISCUSSION TITLE
More attributes for required members

### DISCUSSION THREAD

**Discussion by Pyrdacor** [dotnet/csharplang#6536] [Language Ideas]
At the moment the `required` keyword is used to state that a member has to be initialized. Either via object initializer or by a constructor marked with the `SetsRequiredMembersAttribute`.

This approach is very limited, especially when it comes to derived classes as the attribute only works on constructors of the class which declares the member.

_Example:_

```cs
class Animal
{
    public required int LegCount { get; init; }
}

class Dog : Animal
{
    [SetsRequiredMembers] // not working
    public Dog()
    {
        LegCount = 4;
    }
}
```

In this example the derived class can theoretically provide a value for the required property and initialize it. This is not possible at all with the current implementation.

Even if the attribute would work on derived classes, it would be preferable to have an attribute like `SetsBaseRequiredMembersAttribute` (or `SetsRequiredBaseMembersAttribute`) to not affect all properties of the class itself.

Even better (or maybe in addition) would be an attribute to state that specific required members are initialized. This should then also work on derived classes.

```cs
class Dog : Animal
{
    [SetsRequiredMember(nameof(LegCount))]
    public Dog()
    {
        LegCount = 4;
    }
}
```

This would also be useful to allow constructors which initialize some members.

My proposal would include the following:
- Allow `SetsRequiredMembersAttribute` on derived class constructors. Or change the meaning to "initializes all required members of all base classes as well".
- Add a new attribute to state that a constructor initializes all required members of the base class(es).
- Add a new attribute to state that a constructor initializes a specific required member. This should also work for inherit members, but this should be obvious.

Of course the two new attributes should be allowed both on the same constructor and the single member attribute should be allowed multiple times or should consume a list of member names otherwise.

Any ideas or suggestions?

**Comment by Pyrdacor** [comment]
An optional addition would be to change the contract for a derived class. For example each animal needs to initialize a valid leg count in the example above. But it makes no sense to be able to initialize or change the leg count for a dog. It is an animal and ensures that the leg count has the proper value. So the required member is always initialized, as long as every ctor does it. Ok dogs can lose a leg but there are examples where properties are fixed based on the type you are dealing with. But this might not fit into this proposal.

**Comment by BreyerW** [comment]
I think instead of using completely different attribute for base only members i would use bool parameter instead like `bool BaseOnly=false` or something like that

**Reply by Pyrdacor** [reply]
Good point. I agree.

**Comment by Eli-Black-Work** [comment]
If you can think of one, I'd be interested in seeing more of a real life example 🙂

**Reply by Pyrdacor** [reply]
### Example 1: Parser

```cs
class Expression
{
    public required int TokenCount { get; init; }
    // Other required and non-required properties
}

class IntegerLiteralExpression : Expression
{
    public IntegerLiteralExpression()
    {
        TokenCount = 1; // always
    }
}
```

When parsing I want to do `new IntegerLiteralExpression();` but not `new IntegerLiteralExpression { TokenCount = 10 };`


### Example 2: Game

```cs
class Character
{
    public required Race Race { get; init; }
}

class Monster : Character
{
}

class NPC : Character
{
}

class Player : Character
{
    public Player()
    {
        Race = Race.Human; // player is always human
    }
}
```

You can't even achieve this by workarounds unless you remove the `required` keyword.


### Example 3: Factory

```cs
class Factory<T>
{
    public required Func<T> CreateFunc { get; init; }

    public T Create() => CreateFunc();
}

class MySpecialFactory<T> : Factory<T>
{
    public MySpecialFactory ()
    {
        CreateFunc = () => default(T);
    }
}
```

I think it's common that you define members of a very generic class as required and in more specialized derived classes you can provide some of them based on the context of that derived class. The contract of the derived class no longer necessarily has the same required members then.

**Reply by Pyrdacor** [reply]
It's similar to a constructor. If the base class needs some parameter for the constructor you have to provide it on initialization. But the derived class can provide a parameterless constructor and pass some fixed value to the base class. You can instantiate the derived class without that parameter, even though the base class required it on initialization.

From my perspective a constructor parameter is a required initialization value much like a required property. Even more so if there is only that one constructor. And there you can "remove" that requirement for derived classes with code like this:

```cs
class Character
{
    public Race Race { get; }

    public Character(Race race)
    {
        Race = race;
    }
}

class Player : Character
{
    public Player()
    : base(Race.Human)
    {
    }
}
```

But this approach won't allow something like `var monster = new Monster { Race = Race.Orc };` anymore, even though it would be perfectly fine.

While still I want to use `var player = new Player();`.

**Reply by Pyrdacor** [reply]
With my suggestion it would be:

```cs
class Player : Character
{
    [SetsRequiredMember(nameof(Race))]
    public Player()
    {
        Race = Race.Human;
    }
}
```

You can then use `var player = new Player();` and `var monster= new Monster { Race = Race.Animal };`.

The only caveat here is that you could still do `var player = new Player { Race = Race.Dwarf };`.

Maybe for explicit member attributes there could be an option to forbid this like `[SetsRequiredMember(nameof(Race), allowInitialization = false)]`.

**Reply by Eli-Black-Work** [reply]
@Pyrdacor Thanks, those examples help 🙂

> With my suggestion it would be:
> 
> ```cs
> class Player : Character
> {
>     [SetsRequiredMember(nameof(Race))]
>     public Player()
>     {
>         Race = Race.Human;
>     }
> }
> ```
> 
> You can then use `var player = new Player();` and `var monster= new Monster { Race = Race.Animal };`.
> 
> The only caveat here is that you could still do `var player = new Player { Race = Race.Dwarf };`.
> 
> Maybe for explicit member attributes there could be an option to forbid this like `[SetsRequiredMember(nameof(Race), allowInitialization = false)]`.

I think your last comment provides a pretty good argument for passing `Race` via the constructor, rather than having it be an `init` property.

This feels pretty natural to me:

```c#
class Character
{
    public Race Race { get; }
    
    public Character(Race race)
    {
        Race = race ?? throw new ArgumentNullException(nameof(race));
    }
}

class Monster : Character
{
   public Monster(Race race): base(race)
   {
   }
}

class NPC : Character
{
   public Monster(Race race): base(race)
   {
   }
}

class Player : Character
{
    // player is always human
    public Player() : base(Race.Human)
    {
    }
}
```


In fact, I think that'd be my reply to all of the examples: When inheritance gets involved, it might best to favor constructor parameters over `init`, due to situations like your last comment:

> Maybe for explicit member attributes there could be an option to forbid this like [SetsRequiredMember(nameof(Race), allowInitialization = false)].

**Reply by Eli-Black-Work** [reply]
@Pyrdacor BTW, what's your reason for preferring `init` in these situations?

**Reply by Pyrdacor** [reply]
Well you could argue why you would ever use object initializers then. You can always use constructor arguments.

It may feel natural with only one property but think of some object with many properties. A good example are database entities or DTOs.

The following looks not so clean anymore when all the ctor arguments are necessary, because those are required properties of the base class `Character`.
```cs
var monster = new Monster(Race.Animal, Class.Warrior, 10, true)
{
    SomeNotRequiredProperty = 5
};
```

**Reply by Pyrdacor** [reply]
Object initializers are often more readable.

```cs
var monster = new Monster(true); // bad
var monster = new Monster { CanFly = true }; // better
```

Especially when multiple primitive-type properties like numbers or booleans are involved.

**Reply by spydacarnage** [reply]
```cs
var monster = new Monster(canFly: true):
```
Works just as well?

**Reply by Pyrdacor** [reply]
Sure but arguing like that you can also argue that object initializers make no sense at all and we don't need required properties as we can use constructor parameters.

**Reply by virzak** [reply]
```cs
var monster = new Monster(true);
```
is not bad when you have Visual studio inline hints turned on. You get the best of both worlds: less code and clarity about the parameter.

**Reply by KennethHoff** [reply]
> ```c#
> var monster = new Monster(true);
> ```
> is not bad when you have Visual studio inline hints turned on. You get the best of both worlds: less code and clarity about the parameter.
> 

Late reply (despite reacting to this comment months and months ago). The simple fact that you cannot see the type in the comment you wrote clearly tells us that it is not as clear.

**Comment by Akronae** [comment]
Im also looking for this feature in order to do something like
```csharp
class Animal
{
  public required int LegCount {get; set;}
}

class Bird : Animal
{
  public required float WingSpan {get; set;}

  [SetsRequiredMembers(nameof(LegCount))]
  public Bird(Animal animal)
  {
    LegCount = animal.LegCount;
  }
}

var animal = new Animal { LegCount = 2 };
var bird = new Bird(animal) { WingSpan = 4 };
```

To be honest I'm a bit surprised that we need to use `[SetsRequiredMembers]`, it would be nice if the compiler could infer on its own if all required members are being provided from constructor + object initializer.

**Reply by 333fred** [reply]
> To be honest I'm a bit surprised that we need to use [SetsRequiredMembers], it would be nice if the compiler could infer on its own if all required members are being provided from constructor + object initializer.

If you go back to the [proposal issue](https://github.com/dotnet/csharplang/issues/3630), you'll find plenty of discussion on this topic, including why we wouldn't do that. The tldr is this is required members are part of your public api, and refactorings shouldn't silently change your public api at a distance.

**Reply by FaustVX** [reply]
And also, how deep the compiler should go in the "call stack" and how to deal with branching like
```C#
public C()
{
  MyInitMethod();
  if (_someValue)
    Prop1 = 1;
  else
    Prop2 = 2;
}
```

**Comment by 333fred** [comment]
> especially when it comes to derived classes as the attribute only works on constructors of the class which declares the member. 

I must have missed this comment the first time I read this discussion, as I have no idea what it means. The behavior in derived types is exactly the same as in base types: https://sharplab.io/#v2:C4LglgNgPgAgDAAhgRgHQBEwEMDmA7AewGdgwBjI1AYQIBMBTAQTywgE8iwiBuAWACgBAfQQBeBHnoB3BOgI4AFAEo+g/jABMCZmAC2rAQG8BCU0gDMCAE70AjgFcwN2gjB5gCADL0cNe+4RDBBx6YG5XPDAwhABfATi1TVl5BBBtSP0IIxMzAG0AZVCiACU7R2cAWXpdACN6KyIAXQQAehaJAg8pAisAazccHNMYSzlFJSHAybMvHz8A8QAWVRmEmKA

**Reply by Pyrdacor** [reply]
Strange. Maybe it wasn't working with my version of VS back then? Still the problem remains that you can't distinguish between inherited properties and own properties. In the given example each dog has 4 legs but there might be dog-only properties which should not be set by the ctor.
